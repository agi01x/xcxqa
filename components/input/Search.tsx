import * as React from 'react';
import classNames from 'classnames';
import Input, { InputProps } from './Input';
import Icon from '../icon';
import Button from '../button';

export interface SearchProps extends InputProps {
  inputPrefixCls?: string;
  onSearch?: (value: string, event?: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => any;
  enterButton?: boolean | React.ReactNode;
}

export default class Search extends React.Component<SearchProps, any> {
  static defaultProps = {
    inputPrefixCls: 'ant-input',
    prefixCls: 'ant-input-search',
    enterButton: false,
  };

  private input: Input;

  onSearch = (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => {
    const { onSearch } = this.props;
    if (onSearch) {
      onSearch(this.input.input.value, e);
    }
    this.input.focus();
  }

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  saveInput = (node: Input) => {
    this.input = node;
  }

  getButtonOrIcon() {
    const { enterButton, prefixCls, size, disabled } = this.props;
    const enterButtonAsElement = enterButton as React.ReactElement<any>;
    let node;
    if (!enterButton) {
      node = <Icon className={`${prefixCls}-icon`} type="search" key="searchIcon" />;
    } else if (enterButtonAsElement.type === Button || enterButtonAsElement.type === 'button') {
      node = React.cloneElement(enterButtonAsElement, enterButtonAsElement.type === Button ? {
        className: `${prefixCls}-button`,
        size,
      } : {});
    } else {
      node = (
        <Button
          className={`${prefixCls}-button`}
          type="primary"
          size={size}
          disabled={disabled}
          key="enterButton"
        >
          {enterButton === true ? <Icon type="search" /> : enterButton}
        </Button>
      );
    }
    return React.cloneElement(node, {
      onClick: this.onSearch,
    });
  }

  render() {
    const { className, prefixCls, inputPrefixCls, size, suffix, enterButton, ...others } = this.props;
    delete (others as any).onSearch;
    const buttonOrIcon = this.getButtonOrIcon();
    let searchSuffix = suffix ? [suffix, buttonOrIcon] : buttonOrIcon;
    if (Array.isArray(searchSuffix)) {
      searchSuffix = (searchSuffix as React.ReactElement<any>[]).map((item, index) => {
        if (!React.isValidElement(item) || item.key) {
          return item;
        }
        return React.cloneElement(item, {key: index});
      });
    }
    const inputClassName = classNames(prefixCls, className, {
      [`${prefixCls}-enter-button`]: !!enterButton,
      [`${prefixCls}-${size}`]: !!size,
    });
    return (
      <Input
        onPressEnter={this.onSearch}
        {...others}
        size={size}
        className={inputClassName}
        prefixCls={inputPrefixCls}
        suffix={searchSuffix}
        ref={this.saveInput}
      />
    );
  }
}
