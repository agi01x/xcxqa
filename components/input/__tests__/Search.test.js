import React from 'react';
import { mount } from 'enzyme';
import Search from '../Search';
import Button from '../../button';
import focusTest from '../../../tests/shared/focusTest';

describe('Input.Search', () => {
  focusTest(Search);

  it('should support custom button', () => {
    const wrapper = mount(
      <Search enterButton={<button type="button">ok</button>} />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support custom Button', () => {
    const wrapper = mount(
      <Search enterButton={<Button>ok</Button>} />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support ReactNode suffix without error', () => {
    const fn = () => {
      mount(
        <Search suffix={<div>ok</div>} />
      );
    };
    expect(fn).not.toThrow();
  });

  it('should disable enter button when disabled prop is true', () => {
    const wrapper = mount(
      <Search placeholder="input search text" enterButton disabled />
    );
    expect(wrapper.find('.ant-btn-primary[disabled]')).toHaveLength(1);
  });

  it('should trigger onSearch when click search icon', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" onSearch={onSearch} />
    );
    wrapper.find('.anticon-search').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'click',
      preventDefault: expect.any(Function),
    }));
  });

  it('should trigger onSearch when click search button', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" enterButton onSearch={onSearch} />
    );
    wrapper.find('Button').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'click',
      preventDefault: expect.any(Function),
    }));
  });

  it('should trigger onSearch when click search button with text', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" enterButton="button text" onSearch={onSearch} />
    );
    wrapper.find('Button').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'click',
      preventDefault: expect.any(Function),
    }));
  });

  it('should trigger onSearch when click search button with customize button', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" enterButton={<Button>antd button</Button>} onSearch={onSearch} />
    );
    wrapper.find('Button').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'click',
      preventDefault: expect.any(Function),
    }));
  });

  it('should trigger onSearch when click search button of native', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" enterButton={<button type="button">antd button</button>} onSearch={onSearch} />
    );
    wrapper.find('button').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'click',
      preventDefault: expect.any(Function),
    }));
  });

  it('should trigger onSearch when press enter', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" onSearch={onSearch} />
    );
    wrapper.find('input').simulate('keydown', { key: 'Enter', keyCode: 13 });
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toBeCalledWith('search text', expect.objectContaining({
      type: 'keydown',
      preventDefault: expect.any(Function),
    }));
  });
});
