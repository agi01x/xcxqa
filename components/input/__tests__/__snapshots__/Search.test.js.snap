// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.Search should support custom Button 1`] = `
<span
  class="ant-input-search ant-input-search-enter-button ant-input-affix-wrapper"
>
  <input
    class="ant-input"
    type="text"
  />
  <span
    class="ant-input-suffix"
  >
    <button
      class="ant-btn ant-input-search-button"
      type="button"
    >
      <span>
        ok
      </span>
    </button>
  </span>
</span>
`;

exports[`Input.Search should support custom button 1`] = `
<span
  class="ant-input-search ant-input-search-enter-button ant-input-affix-wrapper"
>
  <input
    class="ant-input"
    type="text"
  />
  <span
    class="ant-input-suffix"
  >
    <button
      type="button"
    >
      ok
    </button>
  </span>
</span>
`;
